langserver/diagnostics.go

```diff
@@ -71,18 +71,31 @@ func (h *LangHandler) publishDiagnostics(ctx context.Context, conn jsonrpc2.JSON
 	return nil
 }
 
-func updateCachedDiagnostics(cachedDiagnostics diagnostics, newDiagnostics diagnostics, files []string) diagnostics {
-	// add/update existing diagnostics
-	for file, diags := range newDiagnostics {
-		cachedDiagnostics[file] = diags
-	}
-
-	// remove all cached diagnostics for each files that is not present in `newDiagnostics`
-	// and add an entry to the output diagnostics for them (to clean the clients)
+func syncCachedDiagnostics(cachedDiagnostics diagnostics, newDiagnostics diagnostics, source string, files []string) diagnostics {
 	for _, file := range files {
-		if _, ok := newDiagnostics[file]; !ok {
-			if _, ok := cachedDiagnostics[file]; ok {
-				delete(cachedDiagnostics, file)
+		_, fileInCache := cachedDiagnostics[file]
+
+		//remove all of the diagnostics for the given source/file combinations and add the new diagnostics to the cache.
+		i := 0
+		for _, diag := range cachedDiagnostics[file] {
+			if diag.Source != source {
+				cachedDiagnostics[file][i] = diag
+				i++
+			}
+		}
+		cachedDiagnostics[file] = append(cachedDiagnostics[file][:i], newDiagnostics[file]...)
+
+		// if the file was already in the cache, the existing diagnostics need sent to the client along with the new
+		if fileInCache {
+			newDiagnostics[file] = cachedDiagnostics[file]
+		}
+
+		// clear out empty cache
+		if len(cachedDiagnostics[file]) == 0 {
+			delete(cachedDiagnostics, file)
+
+			// clear out the client cache
+			if fileInCache {
 				newDiagnostics[file] = nil
 			}
 		}
```

Comment:
so this is unrelated to your code change, but is just so I understand this code again. newDiagnostics is something we send to the client after this function runs. So setting it to nil means we clear it out on the client side.

I wonder if we could simplify all of this and always just send all of cachedDiagnostics (and obviously handle clearing out properly).

Reply:
Notifying the client about diagnostics requires one request per file, so sending all of `cachedDiagnostics` would be a decent amount of extra work.

I do think it could be simplified by having the `diagnostics` that need to be published explicitly returned somewhere instead of `newDiagnostics` being mutated in `syncCachedDiagnostics()`.

I'll spend some time looking at it and seeing if I can clean it up a bit.

Reply:
I've made some changes that I think help this to be slightly more clear.